dopamine/discrete_domains/run_experiment.py

```diff
@@ -607,6 +607,7 @@ def _run_experiment_loop(self):
     iterations is run.
     """
     threads = []
+    self._completed_iteration = self._start_iteration
```

Comment:
If you care enough to redesign and would like help, I'd be happy to go through it together.

I think it's also totally ok to graft this functionality on to Dopamine's design as is.  It maybe adds a little technical debt, but with the project so close to completion, I'm not too concerned.

Reply:
Also if you care but don't have the time to spend on a redesign yourself, let me know and I'll give it a shot.

Reply:
I think i would like for us to find a better implementation at some point. If you want to give it a shot, I would love to see how you approach the problem.
Adding as a TODO(westurner). Is that ok?

Reply:
In term of ideas. There must be some thread-safe increment object that we could pass as an argument to `_run_one_iteration`. That could simply be a queue. Maybe there is something else.

Reply:
> Adding as a TODO(westurner). Is that ok?

Totally.

> In term of ideas. There must be some thread-safe increment object that we could pass as an argument to _run_one_iteration. That could simply be a queue. Maybe there is something else.

That's a good approach.

I wish we could infer that number from the caller based on the number of completed threads.  I don't see any easy way to do that, though.

My own first suggestion is to remove `iteration_number` from all the methods in [Checkpointer](../tree/master/dopamine/discrete_domains/checkpointer.py#L80) and make it a variable tracked internally.  Checkpointer could either surface an `increment_iteration` or make it increment implicitly every time `save_checkpoint` is called.

I don't see a reason that `Runner` should even speak the language of iteration numbers at all (and I kind of wish that `num_iterations` in its interface had been defined in terms of number of iterations to run, not `num_iterations: int, the iteration number threshold (must be greater than start_iteration)`,' which isn't exposed to the caller at all).  In my view, `Runner` only needs to know how many iterations to run.

Reply:
I guess then there's still `_summary_writer` (type `tf.summary.FileWriter`) to take care of.  That takes an iteration for some reason, too.

It'd be a hack, but it could use the iteration number from the checkpointer.  Somewhat better, the checkpointer could also be responsible for writing summaries according to whatever schedule.

If that is done, then `run_one_iteration` doesn't even need to take an iteration number at all.  Which at least to me *feels* right.

Reply:
While we're refactoring, maybe `run_one_iteration` shouldn't be writing summaries or checkpointing anyway.  It could just return the summary, and leave it up to its caller to write them as desired.

Reply:
Well... there's still `_log_experiment`.  Is it bloating the checkpointer to punt the method there, too?

Reply:
```
While we're refactoring, maybe run_one_iteration shouldn't be writing summaries or checkpointing anyway. It could just return the summary, and leave it up to its caller to write them as desired.
```
I think I agree with that.